### ISSUE TITLE
GEM uses memory data in active training set and updates memory twice

### ISSUE THREAD

**Issue by srcdc** [GT-RIPL/Continual-Learning-Benchmark#1]
The `GEM` implementation here inherits from `Naive_Rehearsal` and calls `super(GEM, self).learn_batch(train_loader, val_loader)`. It therefore uses the `learn_batch` method of `Naive_Rehearsal` which uses memory data together with the new data to compute the original gradients before checking conflicts with any memory gradients. From my understanding that is not the intent of the original paper and might affect the training results.

Additionally, `Naive_Rehearal`'s `learn_batch` method already updates the memory and `task_count`, but `GEM` does this a second time once the call returns.

**Comment by yenchanghsu** [comment]
Thanks for pointing out the differences! You are right that the GEM class should not inherit the Naive_Rehearsal class with its current way. I will fix this and check how it impacts the performance.

**Comment by shivamsaboo17** [comment]
hi I was curious to know if this issue is still fixed?

**Comment by srcdc** [comment]
As far as I can tell neither the code nor the results were updated.

**Comment by shivamsaboo17** [comment]
I see. So I made a simple change - added a boolean argument to learn_batch function of NaiveRehearsal using it by passing correction=True: This also prevents repeated update of the memory. Is this implementation correct or have I missed something?
```python
def learn_batch(self, train_loader, val_loader=None, correction=False):
        if correction:
            super(Naive_Rehearsal, self).learn_batch(train_loader, val_loader)
```

**Comment by yenchanghsu** [comment]
Sorry for the late update. The duplicated memory operation in GEM has been removed. Its performance is the same. In the case of split-MNIST-incremental-class, the average accuracy is 92.07+-0.32 (new) versus 92.20+-0.12 (previous result).

I also added the [GEM_orig](https://github.com/GT-RIPL/Continual-Learning-Benchmark/blob/master/agents/customization.py#L174) to match the process in the original GEM paper, which does not include the memory data to calculate the new gradients. The GEM_orig is slightly lower than our implementation (91.32+-0.11 versus 92.07+-0.32).
